### PR TITLE
FE-1677: Enforce optimization constraints and show their verdicts in the study view

### PR DESCRIPTION
**Pull request by kube**
> [!IMPORTANT]
> **Experimental**
> Behind the **In-browser optimization** feature flag.

## Summary

Before this PR, a study recorded its constraints and nothing evaluated them. Every draw the optimizer proposed was simulated, every run counted the same, and the study view had no word on feasibility.

This PR evaluates constraints for a study run in the browser. A parameter constraint is checked before a step simulates, at the trial's values and the net parameter values its scenario resolves them to: a negative margin prunes the step as infeasible with the constraint named, and the step draws grey in the steps table and as a hollow ring on the surface. A state constraint compiles to a 0/1 indicator metric whose minimum over time says whether a run held it throughout, and the trial event carries runs passed over runs total per constraint. The study view gains a Constraints card, a Steps clear stat and a Runs passed column, every rate printed as a raw fraction beside its percentage, and the create drawer gains a Pass threshold. The objective stays the plain mean over every run; verdicts sit beside it and exclude nothing.

https://github.com/user-attachments/assets/ba85d9fa-d333-498b-bb98-4416cb45a45c

## Links

- Ticket [FE-1677](https://linear.app/hash/issue/FE-1677)
- Docs [The Optimizations tab](https://petrinaut-docs-git-claude-opt-proto-constraints.stage.hash.ai/architecture/ui/views/editor/optimizations/optimizations-tab)
- Docs [Optimization › Constraints](https://github.com/hashintel/hash/blob/claude/opt-proto-constraints/libs/@hashintel/petrinaut/docs/optimization.md#constraints)

#### Core

- `constraintMargin` walks a constraint's HIR and gives its signed slack at a point
  > Comparison slack, minimum over `&&`, maximum over `||`, a flipped sign under `!`, the chosen branch of a conditional, scoped `let`, one unit for a bare boolean. `margin >= 0` means satisfied.
  > A zero margin under `!` becomes the boundary failure, so a negated non-strict comparison fails at exact equality.
  > A strict comparison at exact equality returns `-Number.MIN_VALUE`, so the sign is the verdict for integer parameters too.
  > `evaluateParameterConstraints` gives one margin per constraint.
- `compileStateConstraintIndicator` compiles a state constraint as a `body ? 1 : 0` metric
  > Emitted through `emitBufferMetricJs` with no TypeScript compiler, so it runs on the main thread. `interpretHirExpr` is exported for the margin walk.
- HIR interpreter binds locals through a `ReadonlyMap` `Env`
- `getRunValues` honours `aggregateTime` for scalar output
  > `min` over a 0/1 indicator is the "held throughout" quantifier.
- Manifest `constraintPolicy` and trial event `constraints`
  > `constraintPolicy.alpha` strictly between 0 and 1, default 0.05. The trial event carries parameter margins, per-constraint runs passed over runs total, and an optional `infeasible` name. Both trial outcome variants carry the block.
  > The manifest's `constraints` description says the browser runtime evaluates them and reports the results, never enforced on the objective.
- Browser capability merges the channel's results into the trial event
  > Kept per trial on the run record and cleared when the segment ends. The worker and the Python core see nothing of them.

#### Channel

- Parameter constraints checked first, at the trial's resolved net values
  > `parameters.*` is bound to the net parameter values the scenario resolves at the trial's values, overrides applied, through `resolveDetachedObjectiveParameters` on the experiments actions.
  > A negative margin returns `pruned` with `Infeasible: <name>` before `runDetachedObjective` and before `trialStarted`, so no batch appears in the activity list.
- State constraints run as auxiliary metrics of the objective batch
  > `auxiliaryMetrics` on the detached objective request are appended as precompiled scalar expression specs with their `aggregateTime`. Indicators are compiled once per run id and dropped on dispose.
- `trialOutcome` attaches `{ parameters, state }` beside the plain mean
  > A missing run value counts as failed. A constraint the emitter declines throws with its name and the trial is pruned with that reason.
- Followed step's selection stream carries the objective's frames only
  > The batch streams the indicator frames too.

#### Study view

- Constraints card, fourth in a connected study's card row
  > Headline "N / M · P%" over steps clear across the study, subtitle with the pass threshold, alpha and the infeasible draw count, one reserved line for the latest step's verdict, one bar per state constraint with the threshold marked, a note counting parameter constraints.
- Steps clear stat in the summary strip and a Runs passed column in the steps table
  > The column shows the binding constraint's fraction, the lowest pass rate in the step. An infeasible draw's mark is grey, titled "Infeasible: <name>", with greyed row text.
- Infeasible draws are muted rings on the surface and never the best so far
  > `trialFeasibility` reads `constraints.infeasible`; the objective history skips an infeasible draw when accumulating the best so far.
- Pass threshold in the create drawer
  > A percent input under the constraint rows, default 95, shown once a constraint row exists. `constraintPolicyFor` writes `{ alpha }` only when the threshold differs from the default.
- Constraints tooltip and hints in the create drawer say what the browser does with each kind
  > A step that breaks a parameter constraint is skipped before it runs; a run passes a state constraint when the condition holds at every sampled time.
- `constraint-rates.ts` derives every rate from the trials in render
  > `stepVerdict` reads clear, limited, infeasible or unconstrained. `formatRate` prints "52 / 60 · 87%". Nothing is stored.
  > `constraintNameIn` names a constraint for the channel, the card and the table.
- Fixtures and stories for a constrained study
  > `fakeStudyConstraints`, `fakeConstrainedStudyInput`, `makeConstrainedTrials`, and `sirConstrainedOptimizationInput` lowered with the real `lowerConstraint`. Stories `Connected study with constraints` and `Connected study with constraints, following steps`.

#### Review fixes

- `!` fails a constraint whose operand holds exactly at its boundary
  > Negating a zero margin gave `-0`, which reads as satisfied, so `!(x <= y)`, `!(x >= y)` and `!(x == y)` passed at equality.
  > The unary case now maps a zero operand margin to the boundary failure, and `margin.test.ts` pins the three negated cases failing and `!(x < 2)` and `!!(x <= 2)` passing.
- `describeStep` returns the step line as a verdict head and a detail tail
  > The card split the joined string on ` · ` and sliced from `indexOf`, so a parameter-only complete step with no separator rendered "Step N: clearr".
  > The card renders the two parts once, and `constraint-summary.test.ts` and a drawer render test pin the plain "Step N: clear".
- Channel binds `scenario.*` by each scenario parameter's type, a boolean parameter as a boolean
  > It bound the trial's 0/1 transport while `==` compares by identity, so `scenario.flag == true` held for no draw.
  > `decodeScenarioParameterValue` in core is the one rule the scenario compiler and the new `resolveTrialScenarioBindings` apply.
  > `describe.test.ts` pins the decode, and a channel test on a switched SIR scenario expects `scenario.isolation == true` feasible for a true draw and `Infeasible` for a false one.

## Known issues

- Studies on the optimization service report no verdicts
  > The card is mounted for connected studies only.
- Real Pyodide run with a constraint not exercised in this PR
  > The channel and capability tests run constraints lowered from the SIR model and the story carries results through the UI.

## Next steps

- Per-constraint pass rate over steps as a line chart with the threshold marked
- Margins to Optuna
- `AutoStudy` harness `constraints` list for the real Pyodide story

## Test coverage

- `margin.test.ts`, `indicator-metric.test.ts`:
  > Slack per operator, the sign at exact equality, a negated comparison failing at its boundary, `let` scopes, the interpreter's errors, and a compiled indicator that reads 1 then 0 on a two-place net.
- `user-defined.test.ts`, `optimization.test.ts`, `browser-optimization.test.ts`:
  > Per-run minimum over frames, the policy's alpha bounds, the trial event's constraints block, and the capability carrying the channel's results.
- `trial-constraints.test.ts`, `create-optimization-channel.test.ts`, `detached-objective.test.ts`:
  > A pruned infeasible draw before any simulation, a `parameters.*` constraint read at the value the scenario's override resolves, resolved net parameters with the overrides applied, state constraints as auxiliary metrics with per-run verdicts, nothing attached without constraints, auxiliary specs precompiled with their aggregation.
- `constraint-rates.test.ts`, `objective-history-data.test.ts`, `surface-plot.test.tsx`:
  > The threshold rule at equality, the binding constraint, the constraint name lookup, study rates with infeasible draws apart, feasibility read off the constraints block, a muted ring for an infeasible draw.
- `create-optimization-drawer.test.tsx`, `view-optimization-drawer.test.tsx`:
  > A changed threshold written as alpha, the Constraints card, the Steps clear stat, the Runs passed column, greyed marks, a parameter-only study's step line ending at its verdict, and none of it without constraints.
- `constraint-summary.test.ts`:
  > The step line's verdict and detail parts for a parameter-only complete step, a failed step, a binding state constraint and an infeasible draw.
- `describe.test.ts`:
  > A boolean parameter's 0/1 read as the boolean the scenario compiler binds, every other value as it is, and a parameter the values lack read at its default.

## How to test

- Open [Petrinaut preview](https://petrinaut-git-claude-opt-proto-constraints.stage.hash.ai) on Vercel
- Viewport controls > Settings > Simulation > In-browser optimization
- Load example > Supply Chain Profit
- Simulate > Optimizations > Create
- Select metric `Profit`, direction Maximize, tick Optimize production_rate
- Constraints > Add parameter constraint, type `parameters.production_rate <= 320`
- Constraints > Add state constraint, type `return state.places.FinishedGoods.count <= 500;`
  > Expect a Pass threshold field under the rows reading 95
- Run, then open the study row
  > Expect a Constraints card with the steps clear headline and one bar for the state constraint
  > Expect a Steps clear stat in the summary and a Runs passed column in the steps table
- Wait for a draw above 320
  > Expect its row greyed with a mark titled Infeasible and a muted ring on the surface

### CODE REVIEW

File: libs/@hashintel/petrinaut-core/src/constraint/margin.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+  locals: Locals,
+  bindings: HirInterpretBindings,
+): number => {
+  switch (expr.kind) {
+    case "boolLit":
+      return expr.value ? 1 : -1;
+    case "unary":
+      if (expr.op === "!") {
+        return -marginOf(expr.operand, locals, bindings);
+      }
```
[the remark this answers is not in the document]
Comment (kube) [reply libs/@hashintel/petrinaut-core/src/constraint/margin.ts:119]: Fixed in cb8eab2266: `!` now maps a zero operand margin to the boundary failure (`-Number.MIN_VALUE`) instead of negating it to `-0`, so `!(x <= y)`, `!(x >= y)` and `!(x == y)` at exact equality read as failed; margin.test.ts pins the four boundary cases.

File: libs/@hashintel/petrinaut/src/ui/views/Editor/panels/SimulateView/optimizations/study-view/constraint-summary.tsx [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+                className={verdictStyle}
+                data-verdict={stepVerdict(step, alpha)}
+              >
+                {describeStep(optimization, step, alpha).split(" · ")[0]}
+              </span>
+              {describeStep(optimization, step, alpha).slice(
+                describeStep(optimization, step, alpha).indexOf(" · "),
+              )}
+            </>
+          )}
```
[the remark this answers is not in the document]
Comment (kube) [reply libs/@hashintel/petrinaut/src/ui/views/Editor/panels/SimulateView/optimizations/study-view/constraint-summary.tsx:261]: Fixed in cb8eab2266: describeStep now returns the verdict head and the detail tail as separate parts, the card renders them without re-splitting the string, and a test covers the parameter-only complete step that used to render `clearr`.

File: libs/@hashintel/petrinaut/src/react/optimizations/constraint-rates.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+): StepVerdict => {
+  if (!trial.constraints) {
+    return "unconstrained";
+  }
+  if (trial.constraints.infeasible !== undefined) {
+    return "infeasible";
+  }
+  return stepConstraintRates(trial, alpha).every((rate) => rate.passed)
+    ? "clear"
+    : "limited";
```
[the remark this answers is not in the document]
Comment (kube) [reply libs/@hashintel/petrinaut/src/react/optimizations/constraint-rates.ts:85]: The mechanism is right: a completed constrained trial with `runResults.size === 0` gets `state: []` (trial-constraints.ts:97-99), `stepVerdict` reads that as clear (constraint-rates.ts:83-85), and the Constraints card would print a full headline over 0 / 0 rows (constraint-summary.tsx:265-270). It cannot happen in this stack, though. The only backend without a run axis is WebGPU, whose handle keeps `runResults` empty by design (gpu-experiment-handle.ts:617-620), and a browser study cannot run on it: the create drawer disables the Backend switch because the objective is an expression metric the GPU cannot compute (create-optimization-drawer.tsx:800-834; docs/optimization.md:42-47), and a CPU-requested batch pins seeds, which the GPU backend refuses (detached-objective.ts:335-338). The CPU worker posts per-run values for every completed batch (attach.ts:245-250), and a run with a missing indicator counts as failed (trial-constraints.ts:104-107). This becomes a real hole when GPU expression objectives land, so it goes with that work: refuse GPU for a study with state constraints, or prune a constrained step whose batch reported no run axis, with a channel test on sirConstrainedOptimizationInput and a webgpu completion. Filing it as a follow-up rather than changing this PR.

File: libs/@hashintel/petrinaut/src/react/optimizations/channel/create-optimization-channel.ts [hunk 4]
```diff
@@ -82,22 +113,65 @@ export const createOptimizationChannel = ({
...
+          cacheKey,
+          definition: request.manifest.model.definition,
+          scenarioId: request.manifest.scenario.id,
+          scenarioParameterValues: request.scenarioParameterValues,
+          metric: { id: metric.id, label: metric.name, code: metric.code },
+        });
+        parameterOutcome = parameterConstraintOutcome(request.manifest, {
+          parameters,
+          scenario: request.scenarioParameterValues,
+        });
```
[the remark this answers is not in the document]
Comment (kube) [reply libs/@hashintel/petrinaut/src/react/optimizations/channel/create-optimization-channel.ts:142]: Fixed in b994de93de: Fixed in <sha>: confirmed. `resolveTrialScenarioParameterValues` encodes booleans as 0/1 for the scenario compiler (describe.ts:187-188) and the channel bound that transport map as `scenario` (create-optimization-channel.ts:134) while `==`/`!=` compare by identity (margin.ts:52-56); a boolean scenario parameter types as `boolean` in the constraint editor (surface-context.ts:401-410) and `scenario.flag == 1` is rejected as `hir:mixed-equality`, so `scenario.flag == true` was pruned for both values, `scenario.flag != false` passed for both, and `parameters.enabled == scenario.flag` never held (bare `scenario.flag` still worked through truthiness, which is why the numeric SIR tests passed). The channel now binds `scenario` through the decode the scenario compiler already applies to the same values (compile-scenario.ts:263, `type === "boolean" ? value !== 0 : value`), and a channel test with a boolean scenario parameter pins `scenario.isolation == true`: feasible at true, `Infeasible` at false.

**Review by drdma — DISMISSED** [review]

**Review by vilkinsons — APPROVED** [review]
